src/lib/sedlexing.ml

```diff
@@ -216,6 +212,38 @@ let with_tokenizer lexer' lexbuf =
   in
   lexer
 
+module Chan = struct
+  exception Missing_input
+
+  type t = {
+    b : Bytes.t;
+    ic : in_channel;
+    mutable len : int;
+    mutable pos : int;
+  }
+
+  let create ic len : t = { b = Bytes.create len; ic; len = 0; pos = 0 }
+  let available (t : t) = t.len - t.pos
+
+  let rec ensure (t : t) ~can_refill n =
+    if available t >= n then ()
+    else if can_refill then (
+      let len = t.len - t.pos in
+      if len > 0 then Bytes.blit t.b t.pos t.b 0 len;
+      t.pos <- 0;
+      let read = input t.ic t.b len (Bytes.length t.b - len) in
+      if read = 0 then raise Missing_input
+      else (
+        t.len <- len + read;
+        ensure t ~can_refill n))
+    else raise Missing_input
```

Comment:
Is this gonna work if `n` is larger than `len`? You should probably raise an exception if that's not the case.

Reply:
in practice, `n` is 1,2,3 or 4. I can add an assert if you prefer.
I'm not sure what `len` your are referring to.

Reply:
Yes I can see that but for users not familiar with the code, this won't be clear. Adding an assert will make it easier to proof read for any future contributor.

Reply:
Also, in this case why not set the length to 4 instead of parametrizing it?

Reply:
The length of the buffer is (much) bigger than four, allowing to read many bytes at once.

Reply:
I've added some check